src/ansys/additive/core/progress_handler.py

```diff
@@ -19,41 +19,100 @@
 # LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
-"""Provides logging and progress."""
+"""Provides progress handling."""
 
-import logging
+from abc import ABC, abstractmethod
+from enum import IntEnum
 from os import getenv
 
-from ansys.api.additive.v0.additive_domain_pb2 import Progress, ProgressState
+from ansys.api.additive.v0.additive_domain_pb2 import Progress as ProgressMsg
+from ansys.api.additive.v0.additive_domain_pb2 import ProgressState as ProgressMsgState
+from pydantic import BaseModel
 from tqdm import tqdm
 
 
-class ProgressLogger:
-    """Provides for logging and progress reporting."""
+class ProgressState(IntEnum):
+    """Simulation progress state."""
 
-    def __init__(self, name: str = None) -> None:
-        """Initialize a ``ProgressLogger`` object.
+    WAITING = ProgressMsgState.PROGRESS_STATE_WAITING
+    """Waiting for the simulation to start."""
+    RUNNING = ProgressMsgState.PROGRESS_STATE_EXECUTING
+    """Simulation is running."""
+    COMPLETED = ProgressMsgState.PROGRESS_STATE_COMPLETED
+    """Simulation has completed."""
+    ERROR = ProgressMsgState.PROGRESS_STATE_ERROR
+    """Simulation has errored."""
+
+
+class Progress(BaseModel):
+    """Progress information."""
+
+    sim_id: str
+    state: ProgressState
+    percent_complete: int
+    message: str
+    context: str
+
+    @classmethod
+    def from_proto_msg(cls, sim_id: str, progress: ProgressMsg):
+        """Create a ``Progress`` object from a progress protobuf message."""
+        return cls(
+            sim_id=sim_id,
+            state=ProgressState(progress.state),
+            percent_complete=progress.percent_complete,
+            message=progress.message,
+            context=progress.context,
+        )
+
+    def __str__(self):
+        return f"{self.id}: {self.state.name} - {self.percent_complete}% - {self.context} - {self.message}"
+
+
+class IProgressHandler(ABC):
+    """Interface for simulation progress update handlers."""
+
+    @abstractmethod
+    def update(self, progress: Progress):
+        """Update progress.
 
         Parameters
         ----------
-        name: str
-            Name of the logger.
+        progress : Progress
+            Progress information.
         """
-        self._log = logging.getLogger(name)
+        raise NotImplementedError
+
+
+class DefaultSingleSimulationProgressHandler(IProgressHandler):
+    """Creates a :type:`tqdm` progress bar for a single simulation.
+
+    Parameters
+    ----------
+    sim_id : str
+        Simulation ID.
+    """
+
+    def __init__(self):
+        """Initialize progress handler."""
         self._last_percent_complete = 0
         self._last_context = "Initializing"
 
-    def log_progress(self, progress: Progress):
+    def update(self, progress: Progress):
         """Report progress and possibly emit a log message.
```

Comment:
I'm having trouble with "possibly emit." Will a log message be emitted depending on the status? How about: "Report progress and provide a log message as necessary."

Reply:
Good catch! This was a CoPilot generated comment and it's not true. There is no log message emitted.

Reply:
Updated.